t/04-nativecall/02-simple-args.c

```diff
@@ -78,16 +78,16 @@ DLLEXPORT int TakeInt64(int64_t x)
     return 0;
 }
 
-DLLEXPORT int TakeUint8(unsigned char x)
+DLLEXPORT int TakeUint8(char x)
 {
-    if (x == 0xFE)
+    if ((0xFF&(unsigned int)x) == 0xFE)
         return 10;
     return 0;
 }
 
-DLLEXPORT int TakeUint16(unsigned short x)
+DLLEXPORT int TakeUint16(char x)
 {
-    if (x == 0xFFFE)
+    if ((0xFFFF&(unsigned int)x) == 0xFFFE)
```

Comment:
These changes do not make sense to me. We explicitly want to test functions that take unsinged characters and signed shorts. Why would we change both to char or int?

Reply:
@niner the compiler is optimizing out the actual LSB part of the check in clang on arm64.  `w0, w8` are 32bits and actually contain `0xFFFF - 1`.  This test fails even with `-O0` on apple's M1.  Using a 4 byte type and checking explicitly more closely matches what is actually happening when the C is compiled.

The disassembler shows the problem more clearly (this is compiled from the .c prior to my changes).

clang (apple m1, arm64):
```
0000000000003ec4 <_TakeUint8>:
    3ec4: 1f f8 03 71  	cmp	w0, #254   // w0 is 4 bytes and may or may not be 0 padded, eg could be 0x00FE or 0xFFFE

0000000000003ed4 <_TakeUint16>:
    3ed4: c8 ff 9f 52  	mov	w8, #65534  // same problem as with w0.
    3ed8: 1f 00 08 6b  	cmp	w0, w8
```

`%dil` is explicitly checking the LSB, `%di` is checking the 2 LSBs.
gcc (debian on elf64-x86_64):
```
0000000000001260 <TakeUint8>:
    1260:	40 80 ff fe          	cmp    $0xfe,%dil  //explicit LSB check

0000000000001280 <TakeUint16>:
    1280:	66 83 ff fe          	cmp    $0xfffe,%di // explicit 2 LSB check
```

Reply:
Is there any virtue in putting some `#ifdef` so other platforms keep the original behviour?

Reply:
Is there any virtue in putting this in some `#ifdef` so other pl

Reply:
I have not had enough time to investigate this properly. But it feels rather like this is an issue in NativeCall itself. It's quite possible that the platform calling ABI expects us to clear the upper bits of those registers and we don't do that. I'd rather get a definitive answer on what we're expected to do and whether we do this or not, before making any changes to our code.

At least we now know what to look for. That's already a huge step forward on this long standing issue!

Reply:
Hm, I wonder if this is at all related to the errors I'm getting in 03-simple-returns.t with my gmp branch on Windows? A negative int64 is getting corrupted somewhere.

Reply:
@niner the clang implementation says that it can go either way (either the upper bytes are `F|0` but nothing else depending on the signedness coming in) and it's not expected that the upper bits are cleared (even if they aren't this isn't a problem on GCC+x86_64 because of `%dil` rather than the comparison of the entire register).  Either NC can force proper signedness for coercion and try to handle what the compiler _might_ do or we can special case clang on osx to do the right thing.  I noticed the uint16 test is skipped in all cases.  Most likely because they are effectively promoted to `int`s during comparison the value is unsigned -1 for 4|8 bytes. I don't think it's NC's problem to try to handle compiler weirdness.

@MasterDuke17 unless you're using clang this is unrelated.  Use -g3 to create the object and `objdump` can give you more info.

Reply:
If you'd rather use a macro to detect clang on arm to just use int and otherwise let it be, let me know and I'll make it these commits that way

Reply:
> @MasterDuke17 unless you're using clang this is unrelated. Use -g3 to create the object and `objdump` can give you more info.

No, I'm getting the errors with MSVC. Oh well, this sounded vaguely similar/related, but I guess not.